cmd/example-hook-sidecar/smbios.go

```diff
@@ -0,0 +1,114 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"encoding/xml"
+	"net"
+	"os"
+
+	"google.golang.org/grpc"
+
+	vmSchema "kubevirt.io/kubevirt/pkg/api/v1"
+	hooks "kubevirt.io/kubevirt/pkg/hooks"
+	hooksInfo "kubevirt.io/kubevirt/pkg/hooks/info"
+	hooksV1alpha1 "kubevirt.io/kubevirt/pkg/hooks/v1alpha1"
+	"kubevirt.io/kubevirt/pkg/log"
+	domainSchema "kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
+)
+
+const baseBoardManufacturerAnnotation = "smbios.vm.kubevirt.io/baseBoardManufacturer"
+
+type infoServer struct{}
+
+func (s infoServer) Info(ctx context.Context, params *hooksInfo.InfoParams) (*hooksInfo.InfoResult, error) {
+	log.Log.Info("Hook's Info method has been called")
+
+	return &hooksInfo.InfoResult{
+		Name: "smbios",
+		Versions: []string{
+			hooksV1alpha1.Version,
+		},
+		HookPoints: []*hooksInfo.HookPoint{
+			&hooksInfo.HookPoint{
+				Name:     hooksInfo.OnDefineDomainHookPointName,
+				Priority: 0,
+			},
+		},
+	}, nil
+}
+
+type v1alpha1Server struct{}
+
+func (s v1alpha1Server) OnDefineDomain(ctx context.Context, params *hooksV1alpha1.OnDefineDomainParams) (*hooksV1alpha1.OnDefineDomainResult, error) {
+	log.Log.Info("Hook's OnDefinedComain callback method has been called")
+
+	vmJSON := params.GetVm()
+	vmSpec := vmSchema.VirtualMachine{}
+	err := json.Unmarshal(vmJSON, &vmSpec)
+	if err != nil {
+		log.Log.Reason(err).Errorf("Failed to unmarshal given VM spec: %s", vmJSON)
+		panic(err)
+	}
+
+	annotations := vmSpec.GetAnnotations()
+
+	if _, found := annotations[baseBoardManufacturerAnnotation]; !found {
+		log.Log.Info("SM BIOS hook sidecar was requested, but no attributes provided. Returning original domain spec")
+		return &hooksV1alpha1.OnDefineDomainResult{
+			DomainXML: params.GetDomainXML(),
+		}, nil
+	}
+
+	domainXML := params.GetDomainXML()
+	domainSpec := domainSchema.DomainSpec{}
+	err = xml.Unmarshal(domainXML, &domainSpec)
+	if err != nil {
+		log.Log.Reason(err).Errorf("Failed to unmarshal given domain spec: %s", domainXML)
+		panic(err)
+	}
+
+	domainSpec.OS.SMBios = &domainSchema.SMBios{Mode: "sysinfo"}
+
+	if domainSpec.SysInfo == nil {
+		domainSpec.SysInfo = &domainSchema.SysInfo{}
+	}
+	domainSpec.SysInfo.Type = "smbios"
+	if baseBoardManufacturer, found := annotations[baseBoardManufacturerAnnotation]; found {
+		domainSpec.SysInfo.BaseBoard = append(domainSpec.SysInfo.BaseBoard, domainSchema.Entry{
+			Name:  "manufacturer",
+			Value: baseBoardManufacturer,
+		})
+	}
+
+	newDomainXML, err := xml.Marshal(domainSpec)
+	if err != nil {
+		log.Log.Reason(err).Errorf("Failed to marshal updated domain spec: %s", domainSpec)
+		panic(err)
+	}
+
+	log.Log.Info("Successfully updated original domain spec with requested SMBIOS attributes")
+
+	return &hooksV1alpha1.OnDefineDomainResult{
+		DomainXML: newDomainXML,
+	}, nil
+}
+
+func main() {
```

Comment:
Looking at the main() function here, I'm curious if we can abstract away even more involved with the grpc server from the user. It seems like we can make this even simpler for the user. 

Maybe a HookServer interface that would look something like this?

```
// user creates new hook server with a given version, name, and priority
server := hooks.NewHookServer(hooksV1alpha1.Version, "smbios", 0)
defer server.Close()

// register a callback
server.RegisterCallback(hooksV1alpha1.OnDefineDomain, func(ctx context.Context, params *hooksV1alpha1.OnDefineDomainParams) (*hooksV1alpha1.OnDefineDomainResult, error) {
    // add some custom logic for the hook here
}

// begin listening on socket and accepting requests. 
server.Run()
```

That HookServer object would then be able to automatically handle creating/managing the infoServer part on behalf of the user because it would already know the name, version, priority, and callbacks that are registered with the server.

Reply:
Priority must be handled per hook point.

RegisterCallback needs to have static proto buff specification, that would not allow us to change the signature of the hook point or even change it from json-bytes to native struct in case of VM parameter.

I want to add add a nice go-wrapper with the proposed API on top of it. But if you have an idea how to implement it directly on proto buffers, let me know.

Reply:
> RegisterCallback needs to have static proto buff specification, that would not allow us to change the signature of the hook point or even change it from json-bytes to native struct in case of VM parameter.

oh yeah, duh. So it would have to look something like this then. 

```
// user creates a new hook server of a specific version type.
server := hooks.NewHookServerV1Alpha1("smbios")
defer server.Close()

// user registers a callback with priority
server.OnDefineDomainCallback(func(ctx context.Context, params *hooksV1alpha1.OnDefineDomainParams) (*hooksV1alpha1.OnDefineDomainResult, error) {
    // add some custom logic for the hook here
}, 0)

// begin listening on socket and accepting requests. 
server.Run()
```

All this aside, after reviewing the code in more detail I see why you chose your current approach given how the protobuf code is generated.

> I want to add add a nice go-wrapper with the proposed API on top of it. But if you have an idea how to implement it directly on proto buffers, let me know.

I'm for making this a follow up patch. We can discuss improving the client side after you get back from PTO. I agree we should try and get what you've currently done merged first.